core/src/container.cpp

```diff
@@ -790,60 +805,144 @@ bool Alternatives::canCompute() const {
-			active_child_ = nullptr;
-		else
-			active_child_ = next->get();
+	auto impl { pimpl() };
+
+	if (impl->requiredInterface() == GENERATE) {
+		// current_generator_ is fixed if it produced solutions before
+		if (!solutions().empty())
+			return (*impl->current_generator_)->pimpl()->canCompute();
+		else {
+			// move to first generator that can run
+			while(impl->current_generator_ != impl->children().end() && !(*impl->current_generator_)->pimpl()->canCompute()) {
+				ROS_DEBUG_STREAM_NAMED("Fallbacks", "Generator '" << (*impl->current_generator_)->name() << "' can't compute, trying next one.");
+				++impl->current_generator_;
+			}
+			return impl->current_generator_ != impl->children().end();
+		}
 	}
-	return false;
+	else
+		return !impl->pending_states_.empty();
 }
 
 void Fallbacks::compute() {
-	if (!active_child_)
-		return;
+	auto impl { pimpl() };
 
-	try {
-		active_child_->pimpl()->runCompute();
-	} catch (const Property::error& e) {
-		active_child_->reportPropertyError(e);
-	}
+	if(impl->requiredInterface() == GENERATE)
+		impl->computeGenerate();
+	else
+		impl->computeFromExternal();
 }
 
 void Fallbacks::onNewSolution(const SolutionBase& s) {
 	liftSolution(s);
 }
 
+FallbacksPrivate::FallbacksPrivate(Fallbacks* me, const std::string& name)
+   : ParallelContainerBasePrivate(me, name) {
+	current_stage_ = nullptr;
+}
+
+void FallbacksPrivate::initializeExternalInterfaces() {
+	if (requiredInterface() & READS_START)
+		starts().reset(new Interface([this](Interface::iterator external, bool updated) {
+		                  this->onNewExternalState<Interface::FORWARD>(external, updated);
+		               }));
+	if (requiredInterface() & READS_END)
+		ends().reset(new Interface([this](Interface::iterator external, bool updated) {
+		                this->onNewExternalState<Interface::BACKWARD>(external, updated);
+		             }));
+}
+
+void FallbacksPrivate::onNewFailure(const Stage& /*child*/, const InterfaceState* /*from*/, const InterfaceState* /*to*/) {
+	// This override is deliberately empty.
+	// The method prunes solution paths when a child failed to find a valid solution for it,
+	// but in Fallbacks the next child might still yield a successful solution
+	// Thus pruning must only occur once the last child is exhausted (inside computeFromExternal)
+}
+
+void FallbacksPrivate::computeGenerate() {
+	assert(current_generator_ != children().end());
+	(*current_generator_)->pimpl()->runCompute();
+}
+
+template <typename Interface::Direction dir>
+void FallbacksPrivate::onNewExternalState(Interface::iterator external, bool updated) {
+	// TODO(v4hn): updated is not implemented
+	if(updated){
+		ROS_DEBUG_NAMED("Fallbacks", "updating external states is not supported in Fallbacks");
+		return;
+	}
+
+	pending_states_.push_back(ExternalState(external, dir, children().begin()));
+}
+
+void FallbacksPrivate::computeFromExternal(){
+	assert(!pending_states_.empty());
+
+	if(!current_stage_) {
```

Comment:
I'm not sure why you need a `current_stage_`. The current child for each pending external state is stored in `pending_states_` anyway. IMO, you just need to take the ordered list of pending_states_, fetch the most promising first, and try to compute on its child stage. If that fails (and there wasn't a solution computed before), you forward to the next stage but keep this pending state.
Did I miss something? If so, let's have a phone chat.